Connectivity script

jchoude opened the pull request:
This is the first working version of a connectivity script.

It finds the longest GM-GM regions connected by each streamline. Cuts to make sure the streamlines really stay in the regions.

Applies optional post-processing steps: length pruning, loop removal, outlier rejection.

It creates a connectivity matrix, and saves the final bundles. It can optionally save raw connections and intermediate files.

NOTE: there are still a few TODOs that could be improved in future versions.

NOTE: There are TODOs that are related to adding other methods to compute values to assign to the connectivity matrix, such as mean FA, etc. It was decided that this could be added later on in the process.

@mdesco anything else you wanted in this first version?

To be able to use this, you will need to run

```
python setup.py build_ext --inplace
```

mdesco commented:
This is great stuff and usable as is. 

Also, this is where Etienne St-Onge can start for the project with Daniel di Giovanni

mdesco commented:
@jchoude do you have a util script that can easily map a given parcellisation nifti back to integers between 0 and N?

mdesco commented:
Just tested and works very well and is faster than I thought.

mdesco commented:
@jchoude the --verbose option does not do much... Somehow I thought it would tell me which connection is was currently processing... but again, minor

jchoude commented:
For the `--verbose`, I wasn't sure. It had a lot more before, but I didn't know if that was something people really wanted.

For enumerating the currently processed connection, I could add it, but as a debug mode. Else, it will spam your terminal (imagine (95*94) / 2 output lines).

GuillaumeTh commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,351 @@
+    indices, points_to_idx = uncompress(streamlines, return_mapping=True)
+
+    # Compute the connectivity mapping
+    # TODO self connection?
```
Do we remove this comment ?

jchoude replied:
Well it should be implemented someday. Do you prefer creating an issue?

GuillaumeTh commented on scilpy/tractanalysis/tools.py:
```diff
@@ -0,0 +1,170 @@
+             'end_index': end_idx}]
+
+
+# TODO self connect
```
Do we remove this comment ?

jchoude replied:
Same thing.

GuillaumeTh commented on scilpy/tractanalysis/tools.py:
```diff
@@ -0,0 +1,170 @@
+            end_idx = el_idx
+        el_idx -= 1
+
+    # TODO check self connections
```
Do we remove this comment ?

jchoude replied:
Same comment.

GuillaumeTh commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,351 @@
+    con_mat = con_mat[1:, 1:]
+    np.save(os.path.join(args.output, 'final_matrix.npy'), con_mat)
+
+    # In the future, could also be saved as a .json file, using pandas.
```
Do we remove this comment ?

GuillaumeTh commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,351 @@
+Compute a connectivity matrix from a tractogram and a parcellation.
+
+Current strategy is to keep the longest streamline segment connecting
+2 regions. If the streamlines crosses other gray matter regions before
```
the streamlines -> the streamline

jchoude replied:
done

GuillaumeTh commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,351 @@
+    return valid, invalid
+
+
+def build_args_parser():
```
please replace first letter with a capital letter

GuillaumeTh commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,351 @@
+                        'the Nibabel streamlines API.')
+    p.add_argument('labels',
+                   help='labels file name (nifti). Labels must be consecutive '
+                        'from 0 to N, with 0 the background.'
```
Add a space in the end.

GuillaumeTh commented:
Looks good to me ! Everything work fine.

frheault commented:
I need 2-3 functions that are included in that PR, so just in case @mdesco forgot to comment or to confirm it was ok.

mdesco commented:
I approve. Go ahead

On Jun 26, 2019, at 19:09, Francois Rheault <notifications@github.com<mailto:notifications@github.com>> wrote:


I need 2-3 functions that are included in that PR, so just in case @mdesco<https://github.com/mdesco> forgot to comment or to confirm it was ok.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scilus/scilpy/pull/7?email_source=notifications&email_token=AAQSABYBKN4MYQQH2JGSPJDP4OPF7A5CNFSM4HI3LD32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYUGKZY#issuecomment-505963879>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAQSAB3636KKFQ7RJL62G6TP4OPF7ANCNFSM4HI3LD3Q>.

--
Maxime Descoteaux, PhD
Neuroinformatics Research Chair
Sherbrooke Connectivity Imaging Lab (SCIL)
www.scil.usherbrooke.ca<http://www.scil.usherbrooke.ca>

arnaudbore commented:
@jchoude when you have time can you handle these conflicts before merging :)

jchoude commented:
> @jchoude do you have a util script that can easily map a given parcellisation nifti back to integers between 0 and N?

@mdesco I think Gabriel Girard had that once. I doN,t have it directly but I can quickly code it.

jchoude commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,351 @@
+
+This is robust to compressed streamlines.
+
+NOTE: this script can take a while to run. Please be patient.
```
I'll update tomorrow, I need to run said test ;) Also, it highly depends if you save everything or not.

pep8speaks commented:
Hello @jchoude, Thank you for updating !









* In the file [`scripts/scil_compute_connectivity.py`](https://github.com/scilus/scilpy/blob/e60415223accc3f52269f6c8a824e460442257fd/scripts/scil_compute_connectivity.py):

> [Line 119:26](https://github.com/scilus/scilpy/blob/e60415223accc3f52269f6c8a824e460442257fd/scripts/scil_compute_connectivity.py#L119): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'l'
> [Line 292:80](https://github.com/scilus/scilpy/blob/e60415223accc3f52269f6c8a824e460442257fd/scripts/scil_compute_connectivity.py#L292): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 294:80](https://github.com/scilus/scilpy/blob/e60415223accc3f52269f6c8a824e460442257fd/scripts/scil_compute_connectivity.py#L294): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 295:80](https://github.com/scilus/scilpy/blob/e60415223accc3f52269f6c8a824e460442257fd/scripts/scil_compute_connectivity.py#L295): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 296:80](https://github.com/scilus/scilpy/blob/e60415223accc3f52269f6c8a824e460442257fd/scripts/scil_compute_connectivity.py#L296): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 322:80](https://github.com/scilus/scilpy/blob/e60415223accc3f52269f6c8a824e460442257fd/scripts/scil_compute_connectivity.py#L322): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

frheault commented on scilpy/io/streamlines.py:
```diff
@@ -1,9 +1,14 @@
+import six
 
 import nibabel as nib
+from nibabel.streamlines import Field, Tractogram
```
to remove, use dipy.io.streamlines - load_tractogram and save_tractogram instead

frheault commented on scilpy/io/streamlines.py:
```diff
@@ -23,3 +28,63 @@ def check_tracts_same_format(parser, tractogram_1, tractogram_2):
             'Input and output tractogram files must use the same format.')
+
+
+def load_trk_in_voxel_space(trk_file, anat=None, grid_res=None,
```
to remove, use dipy.io.streamlines - load_tractogram and save_tractogram instead

frheault commented on scilpy/io/streamlines.py:
```diff
@@ -23,3 +28,63 @@ def check_tracts_same_format(parser, tractogram_1, tractogram_2):
+    return streamlines
+
+
+def save_from_voxel_space(streamlines, anat, ref_tracts, out_name):
```
to remove, use dipy.io.streamlines - load_tractogram and save_tractogram instead

frheault commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,382 @@
+    p.add_argument('tracks',
+                   help='Path of the tracks file, in a format supported by ' +
+                        'the Nibabel streamlines API.')
+    p.add_argument('labels',
```
Since labels is mandatory, if you use dipy.io.streamlines.load_tractogram this could be your reference

frheault commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,382 @@
+import nibabel as nb
+import numpy as np
+
+from scilpy.io.streamlines import (load_trk_in_voxel_space,
```
Use dipy.io.streamlines.load_tractogram/save_tractogram

frheault commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,382 @@
+def _save_if_needed(streamlines, args, saving_options, out_paths,
+                    save_type, step_type, in_label, out_label):
+    if saving_options[save_type] and len(streamlines):
+        save_from_voxel_space(streamlines, args.labels, args.tracts,
```
if you create a dipy.io.stateful_tractogram.StatefulTractogram and you specify Space.VOX as the last parameter
(you will need to import dipy.io.stateful_tractogram.Space) then the saving can be done with the new dipy API

StongeEtienne replied:
"args.tracts" should be changed to : "args.tracks"
```suggestion
        save_from_voxel_space(streamlines, args.labels, args.tracks,
```

frheault commented on scripts/scil_compute_connectivity.py:
```diff
@@ -0,0 +1,382 @@
+        formatter_class=argparse.RawTextHelpFormatter,
+        description=__doc__)
+    p.add_argument('tracks',
+                   help='Path of the tracks file, in a format supported by ' +
```
To adapt if you switch to StatefulTractogram

GuillaumeTh commented:
@jchoude Do you want to switch to dipy 1.0 as @frheault suggestion ?

jchoude commented:
I updated to master, will finish the last few comments tomorrow.

jchoude commented:
Ok guys, I fixed most of the comments:
- example of runtime added to help
- removed TODOs and created issues

Only missing a script to modify a Freesurfer parc to a contiguous atlas, which should go in another PR, and the migration to StatefulTractogram.

For StatefulTractogram, I have a working version in another branch. However, there are still a few small differences in results that I need to investigate. I know that Guido would like to use this for his OHBM abstract, so if you agree we could merge it like that and then I'll make another PR when my last tests for StatefulTractogram are done.

GuillaumeTh commented:
I'm ok to merge the PR ! Thx @jchoude

GuillaumeTh commented:
@arnaudbore Can I merge?
